README.md

```diff
@@ -102,12 +103,39 @@ In a typical usage scenario you'll want to call one of the `Tracer.getInstance()
 
-**NOTE:** Given the thread-local nature of this library you'll want to make sure the span completion call is in a finally block or otherwise guaranteed to be called no matter what (even if the request fails with an error) to prevent problems when subsequent requests are processed on the same thread. The `Tracer` class does its best to recover from incorrect thread usage scenarios and log information about what happened but the best solution is to prevent the problems from occurring in the first place.
- 
+**NOTE:** Given the thread-local nature of this library you'll want to make sure the span completion call is in a finally block or otherwise guaranteed to be called no matter what (even if the request fails with an error) to prevent problems when subsequent requests are processed on the same thread. The `Tracer` class does its best to recover from incorrect thread usage scenarios and log information about what happened but the best solution is to prevent the problems from occurring in the first place. See the section below on `try-with-resources` for some tips on foolproof ways to safely complete your spans.
+
 <a name="servlet_filter_info"></a>  
 #### Is your application running in a Servlet-based framework?
 
 If your application is running in a Servlet environment (e.g. Spring MVC, Jersey, raw Servlets, etc) then this entire lifecycle can be handled by a Servlet `Filter`. We've created one for you that's ready to drop in and go - see the [wingtips-servlet-api](wingtips-servlet-api/README.md) Wingtips plugin module library for details. That plugin module is also a good resource to see how the code for a production-ready implementation of this library might look.
+
+<a name="try_with_resources_info"></a>
+### Simplified Span Management using Java `try-with-resources` Statements
+
+`Span`s support Java `try-with-resources` statements to help guarantee proper usage in blocking/non-asynchronous scenarios 
+(for asynchronous scenarios please refer to the [asynchronous usage section](#async_usage) of this readme). As 
+previously mentioned, `Span`s that are not properly completed can lead to incorrect distributed tracing information 
+showing up, and the `try-with-resources` statements guarantee that spans are completed appropriately. Here are 
+some examples:
+
+#### Overall request span using `try-with-resources`
+
+``` java
+try(Span requestSpan = Tracer.getInstance().startRequestWith*(...)) {
+    // Traced blocking code for overall request (not asynchronous) goes here ...
+}
```

Comment:
I ran into problems with this design because it doesnt encourage proper error handling and when you add it, you end up with as much code as not using try with resources.

For example, to add response tags or error tags to the span we need to have a catch block and in the finally look if we had a success or exception (or both!)

This and the relative rareness of synchronous api calls that you would trace manually (hence optimize for auto close) pushed towards not doing this mainly to encourage try finally for scoping a span (which is far more routine than finishing on finally)

We may still add on like this (similar to census-java which does the same) but felt should mention.

Reply:
Good points. Since wingtips doesn't currently have a way to signal errors or other tags this is less of a concern at the moment, but it's an issue we'll run into eventually.

Since wingtips encourages fully in-process spans (read: doesn't really support cross-process spans) we actually do use it quite a bit in blocking code, e.g. wrapping a downstream call using a synchronous HTTP client in a subspan.

Reply:
We implemented this pattern in our codebase, and we got more value out of the traces than just "http request starts and ends here", and the syntax in the code just looked good and it stood out more than simple log.info()'s did.

I guess it depends on what you think tracing is for ... is it only intended for automated tracing entry and exit from a sequence of networked resources, or do you want users to explicitly call it for tracing sub-flows within their applications?  (And, for my understanding, are there other use cases you intend for tracing?)

If tracing is intended for only automated entry/exit logging to networked resources, then I'd prefer that the tracing API changes:

* No public API functions.  It would be great to have "configuration only" drive the tracing, wherein middleware takes care of propagating traceId across network and thread boundaries in a flow and logging appropriately.
  * There's no end to async scenarios (custom thread pools, Amazon SMS, etc.), so maybe this is an unrealistic request.  But perhaps the only public API's that would be absolutely necessary would be "saveSpanStackToJson" and "loadSpanStackFromJson" and everything else in the API would be intended for middleware to call and not end-users.
* Explicit instructions in error logs when misconfigurations are detected.  It would be good to specifically direct the developers on what is missing in configurations for systems like hystrix that are preventing traceId from propagating across the thread boundaries and direct the developer how to fix them.  I know there are internal-specifics to these misconfigurations that we would need to explain, so I'm not sure how this language can be delivered in an OSS way.

Reply:
Not all HTTP clients, database drivers, etc, etc, support the capability for auto-instrumenting, so a public API is needed so users can do it themselves without too much hassle. And as you said the async scenarios are many and varied, and the industry is headed more that way every day.

On instructing users when misconfigurations are detected, the potential problems and solutions are again too many and too varied to be able to give much useful advice in the (already pretty verbose) log messages. We could maybe direct them to a troubleshooting section of the readme where we go into more depth on potential problems and solutions?